src/core.js

```diff
@@ -0,0 +1,661 @@
+     */
+
+    /**
+     * @typedef MIDIOutput
+     * @type {MIDIPort}
+     * @property {Function} send - A function that handles requests to send messages to this port.
+     * @property {Function} clear - A function that handles requests to clear messages in progress being sent by this port.
+     */
+
+    youme.requestAccess = function (sysex, software, onAccessGranted, onError) {
+        if (!navigator.requestMIDIAccess) {
+            var msg = "The Web MIDI API is not available. You may need to enable it in your browser's settings.";
+            fluid.log(fluid.logLevel.WARN, msg);
+            onError(msg);
+            return;
+        }
+
+        var p = navigator.requestMIDIAccess({
+            sysex: sysex,
+            software: software
+        });
+
+        p.then(onAccessGranted, onError);
+    };
+
+    youme.getPorts = function (access) {
+        var ports = {};
+
+        youme.collectPorts("inputs", access, ports);
+        youme.collectPorts("outputs", access, ports);
+
+        return ports;
+    };
+
+    youme.requestPorts = function (success, error, software) {
+        function wrappedSuccess(access) {
+            var ports = youme.getPorts(access);
+            success(ports);
+        }
+
+        youme.requestAccess(false, software, wrappedSuccess, error);
+    };
+
+    youme.createPortViews = function (portsArray) {
+        return fluid.transform(portsArray, function (port) {
+            return {
+                id: port.id,
+                name: port.name,
+                manufacturer: port.manufacturer,
+                state: port.state,
+                connection: port.connection
+            };
+        });
+    };
+
+    youme.prettyPrintPorts = function (ports) {
+        return fluid.prettyPrintJSON({
+            inputs: youme.createPortViews(ports.inputs),
+            outputs: youme.createPortViews(ports.outputs)
+        });
+    };
+
+    youme.logPorts = function (software) {
+        function success(ports) {
+            var printed = youme.prettyPrintPorts(ports);
+            console.log(printed); // eslint-disable-line no-console
+        }
+
+        function error(err) {
+            console.log(err); // eslint-disable-line no-console
+        }
+
+        youme.requestPorts(success, error, software);
+    };
+
+    /**
+     * Convert an object that presents a map of `inputs` and `outputs` into an array of MIDIPort objects.
+     *
+     * @param {"inputs"|"outputs"} type - The type of inputs, only "inputs" and "outputs" are supported.
+     * @param {MIDIAccess} access - The `MIDIAccess` object provided by the WebMIDI API.
+     * @param {Map<MIDIPort>} ports - A read-only map-like structure containing the available ports for `type`.
+     * @return {Array<MidiPort>} - An array of `MIDIPort` objects.
+     *
+     */
+    youme.collectPorts = function (type, access, ports) {
+        var portsForType = ports[type] = ports[type] || [],
```

Comment:
Somewhat worrying that each call to this function may end up accumulating further ports in the supplied map - and also that despite being listed as read-only, the ports map is modified. As far as I could see, this effect could be had by simply shallow-cloning "access" into "ports", e.g. 

ports[type] = [...access[type].values()] 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#syntax

Reply:
Done!  On a side note, it's so nice to be able to use ellipses again.

Reply:
Note to self, need to review all map to array usages to confirm whether we would benefit from the spread operator in other places.